### PR TITLE
fixing namespace handling to work as originally intended

### PR DESCRIPTION
**Pull request by jmervine**
You should be able to set `FOO` or `NAMESPACE_FOO` and fetch via `Config.namespace_foo` when doing something like:

```ruby
namespace :namespace do
    optional :foo
end
```

Currently only setting `FOO` gives you a value.

cc @joshuatobin @reidmix

### CODE REVIEW

File: test/configerator_test.rb [hunk 1]
```diff
@@ -20,11 +20,11 @@ class TestConfigerator < Minitest::Test
     test_array_int: [ 9, 9 ],
     test_array: [ 'nine', 'nine' ],
 
-    test_ns1: 'ns1',
-    test_ns2: 'ns2',
+    test_namespace1: 'namespace1',
+    test_namespace2: 'namespace2',
 
-    ns1: 'ns1',
```
Comment (jmervine) [inline test/configerator_test.rb]: @reidmix using `test_ns{1,2}` and `ns{1,2}` is what masked the bug. =P

File: lib/configerator/configerator.rb [hunk 2]
```diff
@@ -76,13 +77,63 @@ def cast(value, method)
...
+    pfxd = prefixize_key(name)
+    meth = pfxd.downcase
 
-    instance_variable_set(:"@#{name}", value)
-    instance_eval "def #{name}; @#{name} || (raise \"key not set '#{name}'\" unless #{error_on_load}) end"
-    instance_eval "def #{name}?; !!#{name} end", __FILE__, __LINE__
+    instance_variable_set(:"@#{meth}", value)
+
+    if has_prefix?
+      instance_eval "def #{meth}; @#{meth} || (raise 'keys not set: \"#{pfxd}\" or \"#{orig}\"' unless #{error_on_load}) end"
```
Comment (jmervine) [inline lib/configerator/configerator.rb]: @reidmix Because only one of the two keys are expected to be set, perhaps this should be `key not set:` and not `keys not set`, in spite of it referencing two keys. Thoughts?

Comment (lexelby) [reply lib/configerator/configerator.rb]: I vote for singular.

File: lib/configerator/configerator.rb [hunk 3]
```diff
@@ -76,13 +77,63 @@ def cast(value, method)
...
+
+  def stringify_key(key)
+    key.to_s.upcase
+  end
+
+  def has_prefix?
+    defined?(@prefix) && !@prefix.nil? && @prefix != ""
+  end
+
+  def prefixize_key(key)
```
Comment (lexelby) [inline lib/configerator/configerator.rb]: IMO the "ize" is unnecessary: "prefix" is already a verb.  `prefix_key` should be sufficient.

File: lib/configerator/configerator.rb [hunk 4]
```diff
@@ -76,13 +77,63 @@ def cast(value, method)
...
+  def fetch_env(key, error_on_load: false, default: nil)
+    stringified_key = stringify_key(key)
+
+    value = if has_prefix?
+              # handle two possible keys
+              prefixed_key = prefixize_key(key)
+
+              ENV[prefixed_key] || ENV[stringified_key] || default
+            else
+              # handle one possible keys
```
Comment (lexelby) [inline lib/configerator/configerator.rb]: s/keys/key/

File: test/configerator_test.rb [hunk 5]
```diff
@@ -63,10 +63,30 @@ def test_required_with_method
...
+  def test_required_missing_on_load
     assert_raises KeyError do
       Config.required :test_missing
     end
+
+    begin
+      Config.required :test_missing
+    rescue => ex
+      assert_equal ex.message, 'key not found: "TEST_MISSING"'
+    end
```
Comment (lexelby) [inline test/configerator_test.rb]: Seems kind of repetitive, having to run `Config.required :test_missing` twice to test both the exception type and message.  Perhaps this?

```
begin
  Config.required :test_missing
rescue KeyError => ex
  assert_equal ex.message, 'key not found: "TEST_MISSING"'
else
  fail
end
```

**Review by lexelby — CHANGES_REQUESTED** [review]

File: test/configerator_test.rb (hunk 5 above)
Comment (jmervine) [reply test/configerator_test.rb]: Oh, that's nifty! Didn't know about `fail`.

**Comment by jmervine** [comment]
@lexelby All comments address. Thanks for the excellent feedback.

File: test/configerator_test.rb [hunk 6]
```diff
@@ -64,14 +64,12 @@ def test_required_with_method
   end
 
   def test_required_missing_on_load
-    assert_raises KeyError do
-      Config.required :test_missing
-    end
-
     begin
       Config.required :test_missing
     rescue => ex
```
Comment (lexelby) [inline test/configerator_test.rb]: Still need to specify `KeyError` here to make sure it raises the right exception.

File: test/configerator_test.rb [hunk 7]
```diff
@@ -168,7 +166,7 @@ def test_namespace_required_missing_on_call
     begin
       Config.test_namespace99
     rescue => ex
-      assert_equal ex.message, 'keys not set: "TEST_NAMESPACE99" or "NAMESPACE99"'
+      assert_equal ex.message, 'key not set: "TEST_NAMESPACE99" or "NAMESPACE99"'
     end
```
Comment (lexelby) [inline test/configerator_test.rb]: Is it possible to do the same begin/rescue/else thing here?

**Review by lexelby — APPROVED** [review]

**Comment by lexelby** [comment]
:+1:

**Comment by jmervine** [comment]
Based on a conversation with @reidmix, I'm going to update this so only `NAMESPACE_FOO` is supported and not `FOO`, when using namespaces by default. Disabling prefix when defining the  namespace will omit the requirement (and support) for the `NAMESPACE_`.

**Comment by jmervine** [comment]
@reidmix Addressed comments from our previous discussion. Switch from minitest to rspec. I found the previous test suite hard to parse and it was leaving me feel less than confident in the tests. Also, because this is a gem, I updated travis to test against all 2.x versions of ruby.

**Comment by reidmix** [comment]
@jmervine awesome on rspec -- yeah i was super lacking on tests plus I'm not always so nimble on minitest.

File: .travis.yml [hunk 8]
```diff
@@ -1,5 +1,9 @@
...
-  - 2.3.1
+  - 2.4.0
+  - 2.3.3
+  - 2.2
+  - 2.1
+  - 2.0
 sudo: false
 bundler_args: --without development
-script: make test
+script: make spec
```
Comment (reidmix) [inline .travis.yml]: is it weird we're using `make`?

File: lib/configerator/configerator.rb [hunk 9]
```diff
@@ -69,20 +70,39 @@ def array(method = nil)
...
     instance_eval "def #{name}?; !!#{name} end", __FILE__, __LINE__
 
+    return unless @is_namespace
+
     @processed ||= []
     @processed << name
   end
+
+  def stringify_key(key)
+    key = "#{@prefix}#{key}" if defined?(@prefix) && !@prefix.nil? && @prefix != ""
```
Comment (reidmix) [inline lib/configerator/configerator.rb]: I believe you can safely do `if @prefix` to simplify and get what you're looking for. 

It'll return `nil` if its undefined.
If it's an empty string, it won't change the `key`.

File: lib/configerator/configerator.rb [hunk 10]
```diff
@@ -69,20 +70,39 @@ def array(method = nil)
...
+    instance_eval "def #{name}; @#{name} || (raise 'key not set: \"#{name.upcase}\"' unless #{error_on_load}) end", __FILE__, __LINE__
     instance_eval "def #{name}?; !!#{name} end", __FILE__, __LINE__
 
+    return unless @is_namespace
+
     @processed ||= []
     @processed << name
   end
+
+  def stringify_key(key)
```
Comment (reidmix) [inline lib/configerator/configerator.rb]: I'd change the name here because its doing more than stringifying.  maybe `build_key`?

File: lib/configerator/configerator.rb [hunk 11]
```diff
@@ -69,20 +70,39 @@ def array(method = nil)
...
+    key = "#{@prefix}#{key}" if defined?(@prefix) && !@prefix.nil? && @prefix != ""
+
+    key.to_s.upcase
+  end
+
+  def fetch_env(key, error_on_load: false, default: nil)
```
Comment (reidmix) [inline lib/configerator/configerator.rb]: I don't think the `default` handling should be in this method, primarily because it's for one usecase (using `override`) and pushing it into this method complicates this method in subtle ways. 

The `error_on_load` functionality is no longer being executed quite the same way.  It's what `ENV.fetch` was doing when it raised an error if the key didn't exist, which is different from the value being `nil` (I'm realizing your change was done to help handle the case of applying the `default` values). 

My suggestion is to move it back to using `ENV.fetch` and let it do the error raising and not raise the error ourself here.   Then in `override` apply the default there (I'll make a note of it.)

File: lib/configerator/configerator.rb [hunk 12]
```diff
@@ -1,36 +1,37 @@
...
   end
 
   def optional(name, method=nil)
-    value = cast(ENV[name.to_s.upcase], method)
+    value = cast(fetch_env(name), method)
+    create(name, value)
+  end
+
+  def override(name, default, method=nil)
+    value = cast(fetch_env(name, default: default), method)
```
Comment (reidmix) [inline lib/configerator/configerator.rb]: See my longer note on `fetch_env` -- I think we can simplify `fetch_env` if we apply the default inline here:

```ruby
    value = cast(fetch_env(name) || default, method)
```

File: lib/configerator/configerator.rb [hunk 13]
```diff
@@ -1,36 +1,37 @@
...
 
 module Configerator
-  @processed = []
-
+  # Initializers (DSL)
   def required(name, method=nil, error_on_load: true)
-    # Hash#fetch raises a KeyError, Hash#[] doesn't
-    value = error_on_load ? ENV.fetch(name.to_s.upcase) : ENV[name.to_s.upcase]
-    value = cast(value, method)
+    value = fetch_env(name, error_on_load: error_on_load)
```
Comment (reidmix) [inline lib/configerator/configerator.rb]: I'd probably make this look more like `optional` & `override` when it gets the value, for consistency:
```ruby
value = cast(fetch_env(name, error_on_load: error_on_load), method)
```

File: lib/configerator/configerator.rb [hunk 14]
```diff
@@ -69,20 +70,39 @@ def array(method = nil)
...
-    name = "#{@prefix}#{name}"
+    name = stringify_key(name).downcase
 
     instance_variable_set(:"@#{name}", value)
-    instance_eval "def #{name}; @#{name} || (raise \"key not set '#{name}'\" unless #{error_on_load}) end"
+
+    instance_eval "def #{name}; @#{name} || (raise 'key not set: \"#{name.upcase}\"' unless #{error_on_load}) end", __FILE__, __LINE__
     instance_eval "def #{name}?; !!#{name} end", __FILE__, __LINE__
 
+    return unless @is_namespace
```
Comment (reidmix) [inline lib/configerator/configerator.rb]: I don't think it hurts to have `@processed` get created if its not a namespace -- what you have here to check if it's for `@is_namespace` is more correct (for sure!) but I don't think you really need it.  the `@is_namespace` check doesn't really offer anything.  `@processed` gets cleared out every time `namespace` is called, so it stays tidy. 

Personally, I wouldn't worry about it and get rid of `@is_namespace`

File: spec/configerator_spec.rb [hunk 15]
```diff
@@ -0,0 +1,275 @@
+require 'uri'
+require 'spec_helper'
+
+describe Configerator do
+  let(:config) { Module.new { extend Configerator } }
```
Comment (reidmix) [inline spec/configerator_spec.rb:5]: sweet!

File: spec/configerator_spec.rb [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+    it "loads required value" do
+      with_environment foo: "bar" do
+        config.required :foo
+
+        expect(config.foo).to eq "bar"
+        expect(config.foo?).to be true
+      end
+    end
+
+    it "errors when missing required value" do
```
Comment (reidmix) [inline spec/configerator_spec.rb:17]: I'd also do a test for when `FOO` is not in the environment at all.

File: spec/configerator_spec.rb [hunk 17]
```diff
@@ -0,0 +1,275 @@
...
+    it "loads optional value" do
+      with_environment foo: "bar" do
+        config.optional :foo
+
+        expect(config.foo).to  eq "bar"
+        expect(config.foo?).to be true
+      end
+    end
+
+    it "accepts missing value" do
```
Comment (reidmix) [inline spec/configerator_spec.rb:51]: I'd also do a test for when FOO is not in the environment at all.

File: spec/configerator_spec.rb [hunk 18]
```diff
@@ -0,0 +1,275 @@
...
+      with_environment foo: "bar" do
+        config.override :foo, "bah"
+
+        expect(config.foo).to  eq "bar"
+        expect(config.foo?).to be true
+      end
+    end
+
+    it "uses default with missing value" do
+      with_environment foo: nil do
```
Comment (reidmix) [inline spec/configerator_spec.rb]: I'd also do a test for when FOO is not in the environment at all.

File: spec/configerator_spec.rb [hunk 19]
```diff
@@ -0,0 +1,275 @@
...
+        end
+      end
+
+      it "applies method when not set" do
+        with_environment foo: nil do
+          config.override :foo, "bah", method
+
+          expect(config.foo).to  eq "method:bah"
+        end
+      end
```
Comment (reidmix) [inline spec/configerator_spec.rb]: 👍

File: spec/spec_helper.rb [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+  allow(ENV).to  receive(:[]).and_call_original
+end
+
+def with_environment env
+  stub_environment(env)
+
+  yield
+
+  unstub_environment
+end
```
Comment (reidmix) [inline spec/spec_helper.rb:21]: I really like how this turned out in the tests, it looks really nice 👍

**Review by reidmix — CHANGES_REQUESTED** [review]
This is looking really good -- I really like what you did with the tests. 

Most of my comments are around simplifying the code more than anything.

File: spec/configerator_spec.rb (hunk 16 above)
Comment (jmervine) [reply spec/configerator_spec.rb:17]: Setting it to `nil` is how you emulate it not being in the environment. If I don't remember it all all the `allow_any_instance_of` throws an error b/c something is called that's not expected. I might be able to do that by not stubbing the environment for that test, but then if someone happens to have `FOO` set, it will break the spec.

File: lib/configerator/configerator.rb (hunk 11 above)
Comment (jmervine) [reply lib/configerator/configerator.rb]: Yeah, I was on the fence on this, I'll revisit.

File: .travis.yml (hunk 8 above)
Comment (jmervine) [reply .travis.yml]: Yeah, it tends to be my goto for simple things, I'll switch it to rake at some point, but perhaps not in this PR.

File: lib/configerator/configerator.rb (hunk 9 above)
Comment (jmervine) [reply lib/configerator/configerator.rb]: I was trying to catch the `""` condition as well, but I think that's overkill. Good call here.

File: spec/configerator_spec.rb (hunk 16 above)
Comment (reidmix) [reply spec/configerator_spec.rb:17]: what if you don't mock anything?  that should do it right?

Comment (jmervine) [reply spec/configerator_spec.rb:17]: Yeah, oddly, initially I was getting errors with that, but I think that was in a previous iteration. I'm not sure what I changed to make it work, but that's what I'm doing now and it looks good.

File: lib/configerator/configerator.rb [hunk 21]
```diff
@@ -69,20 +69,34 @@ def array(method = nil)
...
+
+    instance_eval "def #{name}; @#{name} || (raise 'key not set: \"#{name.upcase}\"' unless #{error_on_load}) end", __FILE__, __LINE__
     instance_eval "def #{name}?; !!#{name} end", __FILE__, __LINE__
 
     @processed ||= []
     @processed << name
   end
+
+  def build_key(key)
+    key = "#{@prefix}#{key}" if @prefix
```
Comment (reidmix) [inline lib/configerator/configerator.rb:92]: i'm realizing we don't even need the `if @prefix` because if it's nil or blank, we'd never see it. 
Something to simplify for the future, not required to change now.

**Review by reidmix — APPROVED** [review]
LGTM 👍
